### ISSUE TITLE
[RFC]: Unified Diffusion MoE LoRA via MoERunner

### ISSUE THREAD

**Issue by BLANKETusers** [vllm-project/vllm-omni#8098]
### Motivation.

vLLM-Omni 已支持 diffusion Dense Linear LoRA。对于 diffusion MoE，我们不应该为每个模型实现独立的 Expert LoRA runtime，而应该将routed experts 统一接入 vLLM`MoERunner`，并复用上游 `FusedMoEWithLoRA` 及其平台实现。

当前 diffusion 模型可以分成三类：

| 类型 | 代表模型 | Base MoE | LoRA 路径 |
| --- | --- | --- | --- |
| 直接使用 `MoERunner` | HunyuanImage3 | 上游原生支持 | `FusedMoEWithLoRA` |
| 通过 compatibility patch 使用 `MoERunner` | Moda2.5、Moda3、已接入的 LingBotVideo、HiDreamImage | 补齐 NPU non-gated capability | `FusedMoEWithLoRA` |
| 无 MoE 的全 Dense 模型 | Wan2.2、Qwen-Image | 不涉及 | Diffusion Linear LoRA |

前两类在 LoRA 架构上是同一类：

```text
DiffusionLoRAManager
  -> MoERunner
  -> vLLM FusedMoEWithLoRA
  -> platform MoE LoRA implementation
```

NPU 平台使用
`vllm_ascend.lora.fused_moe.AscendFusedMoEWithLoRA`。该实现继承
`FusedMoEWithLoRA`，复用上游 LoRA weight allocation、`set_lora()`、
`reset_lora()`、TP/EP slicing 和 `MoELoRAContext`。Ascend 只负责平台相关的LoRA 注入。

Moda2.5/3 的 compatibility patch 只补齐 base `MoERunner` 的 NPU non-gated capability，不引入另一套 MoE LoRA。

### Proposed Change.

所有 token-routed diffusion MoE 应统一使用：

```text
Diffusion model
  -> vllm_omni.diffusion.layers.FusedMoE
  -> vLLM MoERunner
  -> FusedMoEWithLoRA
  -> platform MoE LoRA
```

模型侧负责描述 expert、router、checkpoint 和 shared expert 语义。框架侧负责
`MoERunner` discovery、LoRA checkpoint 加载和转换、wrapper replacement、adapter
生命周期以及平台执行。

本 RFC 不增加 `SupportsExpertLoRA`、`prepare_lora_weights()` 或模型私有
`set_lora()/reset_lora()` 协议。尚未使用 `MoERunner` 的 token-routed MoE 应先迁移
runner，而不是增加长期维护的第二套 LoRA compute。

### Current model paths.

HunyuanImage3 是直接接入 `MoERunner` 的基线。它使用标准 gated experts：

```text
gate, up = split(x @ W13)
hidden = activation(gate) * up
output = hidden @ W2
```

模型直接构造 diffusion `FusedMoE`：

```python
self.experts = FusedMoE(
    shared_experts=self.shared_mlp,
    num_experts=self.n_routed_experts,
    top_k=top_k,
    hidden_size=config.hidden_size,
    intermediate_size=intermediate_size,
    renormalize=top_k > 1,
    quant_config=quant_config,
    prefix=f"{prefix}.experts",
)
```

forward 只产生 router logits 并调用 runner：

```python
router_logits, _ = self.gate(hidden_states)
output = self.experts(
    hidden_states=hidden_states,
    router_logits=router_logits,
)
```

基础 checkpoint 通过上游 helper 映射到 `w13/w2`：

```python
FusedMoE.make_expert_params_mapping(
    ckpt_gate_proj_name="gate_proj",
    ckpt_down_proj_name="down_proj",
    ckpt_up_proj_name="up_proj",
    num_experts=config.num_experts,
)
```

Moda2.5/3 的 expert 语义当前不能由 Ascend `MoERunner` 直接执行，因此需要
compatibility patch。patch 后构造结果仍然是上游 `MoERunner`，MoE LoRA 仍然进入
`FusedMoEWithLoRA`。

Wan2.2、Qwen-Image 等全 Dense 模型没有 routed experts，继续使用现有 Diffusion
Linear LoRA：

```text
output = Linear(x, W) + (x @ A) @ B
```

### Why Moda2.5/3 require a compatibility patch.

Moda2.5/3 的配置明确指定：

```python
gated_linear_unit = False
```

其 expert 是两投影 non-gated GELU：

```text
hidden = GELU_TANH(x @ weight1)
output = hidden @ weight2
```

以 Moda3 为例：

```text
hidden_size = 3072
num_experts = 128
intermediate_size = 1024

weight1 = [3072, 128 * 1024] = [3072, 131072]
weight2 = [128 * 1024, 3072] = [131072, 3072]
```

如果是 gated FFN，第一层需要 gate/up 两套投影：

```text
[3072, 128 * 1024 * 2] = [3072, 262144]
```

因此 Moda 必须使用：

```python
activation = MoEActivation.GELU_TANH_NO_MUL
is_act_and_mul = False
```

vLLM 虽然定义了 `GELU_TANH_NO_MUL`：

```python
class MoEActivation(Enum):
    GELU_TANH_NO_MUL = "gelu_tanh_no_mul"

    @property
    def is_gated(self) -> bool:
        return not self.value.endswith("_no_mul")
```

但 `FusedMoEConfig` 在 NPU 平台拒绝 `is_act_and_mul=False`：

```python
if not self.is_act_and_mul and not (
    current_platform.is_cuda_alike()
    or current_platform.is_xpu()
):
    raise NotImplementedError(
        "is_act_and_mul=False is supported only for "
        "CUDA, XPU and ROCm for now"
    )
```

Moda 会在 runner 构造阶段失败。即使绕过该检查，当前 vLLM-Ascend unquantized MLP
也只显式处理 gated GELU：

```python
if activation == MoEActivation.GELU:
    gate, up = gate_up_out.chunk(2, dim=-1)
    gate_up_out = F.gelu(gate) * up
elif activation == MoEActivation.GELU_TANH:
    gate, up = gate_up_out.chunk(2, dim=-1)
    gate_up_out = F.gelu(gate, approximate="tanh") * up
else:
    gate_up_out = torch_npu.npu_swiglu(gate_up_out)
```

缺少 Moda 需要的单投影分支：

```python
if activation == MoEActivation.GELU_TANH_NO_MUL:
    gate_up_out = F.gelu(
        gate_up_out,
        approximate="tanh",
    )
```

因此 Moda2.5/3 不能直接接入当前 Ascend `MoERunner` 的原因是：

1. vLLM platform check 不允许 NPU 使用 `is_act_and_mul=False`。
2. vLLM-Ascend 缺少 `GELU_TANH_NO_MUL` activation。
3. Moda checkpoint 使用 packed `weight1/weight2`，runner 参数为 `w13/w2`。
4. Moda router 需要 FP32 logits、sigmoid、expert correction bias 和特定归一化顺序。

前两项由通用 platform capability patch 解决，后两项由模型 loader 和
`custom_routing_function` 表达。

### MoERunner model integration.

以 Moda2.5/3 为例，接入工作分为模型侧和框架侧两部分。

#### Model-side work.

1. **构造 `MoERunner`。**

   模型使用 diffusion `FusedMoE` 描述 expert shape、activation 和并行配置：

   ```python
   self.experts = FusedMoE(
       num_experts=num_experts,
       top_k=top_k,
       hidden_size=hidden_size,
       intermediate_size=intermediate_size,
       activation="gelu_tanh_no_mul",
       scoring_func="sigmoid",
       renormalize=True,
       custom_routing_function=self._select_experts,
       router_logits_dtype=torch.float32,
       quant_config=None,
       prefix=f"{prefix}.experts",
   )
   ```

   `FusedMoE.__new__()` 返回真实 `MoERunner`，使 runner 参数可以被
   `named_parameters()`、DLO 和 checkpoint loader 发现。

2. **保留 router 语义。**

   Moda 使用 custom routing function 保留 FP32、sigmoid、expert bias 和归一化顺序：

   ```python
   router_scores = torch.sigmoid(gating_output)
   selection_scores = router_scores + self.router.expert_bias.float()
   expert_indices = torch.topk(selection_scores, topk, dim=-1).indices
   expert_weights = torch.gather(
       router_scores,
       -1,
       expert_indices,
   )
   expert_weights /= expert_weights.sum(dim=-1, keepdim=True) + 1e-20
   ```

   `expert_bias` 只影响专家选择，不进入最终 combine weight。

3. **映射 base checkpoint。**

   模型 loader 转换 packed 权重，再逐 expert 调用上游 `weight_loader`：

   ```text
   weight1 [H, E*I] -> w13_weight [E, I, H]
   weight2 [E*I, H] -> w2_weight  [E, H, I]
   ```

   EP/TP slicing 继续由 runner expert loader 负责。

4. **声明 LoRA target。**

   模型只声明 checkpoint 中的逻辑 target：

   ```python
   packed_modules_mapping = {
       "experts": ["weight1", "weight2"],
   }
   ```

   模型不实现 LoRA GEMM，不管理 adapter cache，也不保存 active LoRA tensor。

5. **明确 shared expert 边界。**

   Moda 的 shared expert 保持在 routed `MoERunner` 外计算。若 shared expert 需要 LoRA，
   应按 Dense Linear LoRA 单独接入。

#### Framework-side work.

1. **补齐 base MoE capability。**

   Omni 在创建 NPU unquantized runner 时使用 scoped compatibility context：

   ```python
   with allow_unquantized_no_mul_fused_moe():
       moe_runner = FusedMoE(prefix=prefix, **kwargs)
   ```

   patch 仅在 `quant_config=None` 时跳过 NPU non-gated 配置拒绝，并增加
   `GELU_TANH_NO_MUL`，不修改 gated 和 quantized MoE。

2. **发现并包装 `MoERunner`。**

   manager 根据模块类型和模型 metadata 发现 runner，在 NPU 上选择 Ascend wrapper：

   ```python
   if isinstance(layer, MoERunner):
       wrapper_cls = FusedMoEWithLoRA
       if current_platform.device_type == "npu":
           wrapper_cls = AscendFusedMoEWithLoRA
       lora_layer = wrapper_cls(layer)
   ```

   Moda、HunyuanImage3 和后续 `MoERunner` 模型共享该 wrapper selection。

3. **加载并转换 LoRA checkpoint。**

   Moda `weight1/weight2` LoRA 被转换为：

   ```text
   weight1: [local experts, rank, hidden]
            [local experts, intermediate, rank]

   weight2: [local experts, rank, intermediate]
            [local experts, hidden, rank]
   ```

   manager 在 CPU 上完成 shape normalization，按 `ep_rank` 保留 local experts，
   并将请求级 scale 乘到 B。

   上游 `set_lora()` 当前仍按 `[w1, w2, w3]` 解包，但 non-gated runner 只分配一个
   `w13` slice：

   ```python
   self._w13_slices = (
       2 if base_layer.moe_config.is_act_and_mul else 1
   )
   ```

   Moda 传入的第三个 `w3` 只是 API 兼容占位，不会被复制或参与计算：

   ```python
   lora_layer.set_lora(
       index=0,
       lora_a=[weight1_a, weight2_a, weight1_a],
       lora_b=[weight1_b, weight2_b, weight1_b],
   )
   ```

4. **复用上游 MoE LoRA runtime。**

   框架直接复用：

   - `FusedMoEWithLoRA.create_lora_weights()`；
   - `FusedMoEWithLoRA.set_lora()/reset_lora()`；
   - `MoELoRAContext`；
   - Punica token-to-adapter mapping；
   - vLLM-Ascend `moe_lora_apply_w13()` 和 `moe_lora_apply_w2()`。

5. **管理 runtime context 和 adapter 生命周期。**

   框架负责 adapter ID、cache、LRU、scale、activate、switch、reset 和 unload。

   上游 `FusedMoEWithLoRA.forward()` 直接调用 `base_layer.forward()`，会绕过原 runner
   module hook，因此框架需要将 pre-hook 复制到 wrapper，继续初始化 token count、
   DP/EP routing metadata、Ascend dispatch context 和 DLO execution device。

完成这些通用框架工作后，HunyuanImage3 等直接使用 `MoERunner` 的模型原则上只需在
模型侧提供 checkpoint target mapping 和 shared expert 策略。只有新模型引入现有平台
无法表达的 activation、量化或通信方式时，才需要增加框架能力。

### Current framework gaps.

当前 manager 的 target discovery 仍硬编码 Moda mapping：

```python
if isinstance(module, MoERunner):
    packed = self._packed_modules_mapping.get("experts")
    if packed == ["weight1", "weight2"]:
        return list(packed)
```

需要将其泛化为模型声明的任意 MoE target metadata，覆盖 HunyuanImage3 的
`gate_proj/down_proj/up_proj`。

当前 Ascend wrapper 还会拒绝带 shared experts 的 runner：

```python
if getattr(base_layer, "_shared_experts", None) is not None:
    raise AssertionError(
        "Ascend MoE LoRA does not wrap the shared_experts path"
    )
```

HunyuanImage3 会把 `shared_mlp` 传入 runner。可选择只对 routed experts 应用 MoE
LoRA，或者将 shared expert 从 runner 解耦并单独使用 Dense Linear LoRA。

### Correctness and testing.

已完成：

1. Moda2.5/3 base MoE 迁移到 `MoERunner`。
2. NPU unquantized `gelu_tanh_no_mul` 与 PyTorch reference parity。
3. Moda base checkpoint `weight1/weight2 -> w13/w2` 映射。
4. FP32 router、sigmoid、expert bias、top-k 和 combine weight parity。
5. `MoERunner -> AscendFusedMoEWithLoRA` wrapper replacement。
6. Moda LoRA A/B normalization、EP local slicing 和 scale。
7. Punica token mapping、runner pre-hook 和 execution device。
8. 单卡及 2 卡 EP synthetic MoE LoRA parity。
9. Moda2.5/3 DP4/EP4/DLO base model smoke。

生产验收前还需要：

1. 使用真实 Moda LoRA checkpoint 完成单卡 base/LoRA 视频对比。
2. 验证真实 adapter 的 DP/EP + LoRA 和 DP/EP + DLO + LoRA。
3. 覆盖 adapter load、scale、switch、reset、unload 和 LRU。
4. 验证 request batch 中不同 adapter 请求的一致性。
5. 对比 router indices、weights、expert 输出和 DiT prediction。
6. 使用固定输入、步数、并发和 backend 做迁移前后性能 A/B。
7. 为 HunyuanImage3 补齐 target mapping 和 shared expert 支持后执行同类验证。
8. 回归 Dense Linear LoRA，确保 MoE bridge 不改变原有行为。

当前环境未找到真实 Moda LoRA adapter。因此现阶段结论是：Moda3 已采用 patched
`MoERunner + FusedMoEWithLoRA`，代码链路和 synthetic parity 已验证，真实 adapter
的完整生产精度尚未验收。

### Risks.

| Risk | Impact | Mitigation |
| --- | --- | --- |
| checkpoint target 与 runner target 不一致 | adapter 未加载或 projection 错位 | 模型显式声明 mapping，并校验名称和 shape |
| non-gated `w3` 兼容参数被误用 | Moda 数值错误 | 校验 `is_act_and_mul=False` 和 `_w13_slices == 1` |
| EP rank 缓存完整 adapter | CPU/NPU 内存浪费 | normalization 后只保留 local experts |
| wrapper 绕过 runner pre-hook | routing metadata 未初始化 | 将 pre-hook 复制到 wrapper |
| DLO 根据 base weight 推断错误 device | LoRA buffer 落在 CPU/meta | 显式传入 execution device |
| compatibility patch 影响其他 MoE | gated 或量化模型回归 | patch 限定 NPU、no-mul、`quant_config=None` |
| Ascend shared expert 限制 | Hunyuan wrapper 初始化失败 | routed LoRA 与 shared Dense LoRA 分离 |
| 只有 synthetic adapter 验证 | 真实 checkpoint 命名或质量问题 | 合入前增加真实 adapter E2E |
| custom 旧代码继续存在 | 维护者误判实际路径 | 完成回归后删除 dead implementation |

### Open questions.

1. MoE target mapping 是否统一定义为模型 metadata，并同时服务 base loader 和 LoRA
   loader？
2. HunyuanImage3 shared expert 是保持 base-only，还是单独接入 Dense Linear LoRA？
3. 上游 `set_lora()` 是否应允许 non-gated 模型只传 `[w1, w2]`？
4. quantized non-gated MoE 是否纳入下一阶段？

## Implementation Plan

This work will be split into a few PRs:

1. Generic MoE LoRA Bridge for vLLM-Omni.
2. Moda2.5/3 MoERunner Integration.
3. Upstream vLLM and vLLM-Ascend Capability Support.
4. HunyuanImage3 MoE LoRA Integration.

### References.

- vLLM `FusedMoEWithLoRA`:
  `vllm/lora/layers/fused_moe.py`
- vLLM non-gated platform check:
  `vllm/model_executor/layers/fused_moe/config.py`
- vLLM MoE activation definitions:
  `vllm/model_executor/layers/fused_moe/activation.py`
- vLLM-Ascend MoE LoRA:
  `vllm_ascend/lora/fused_moe.py`
- vLLM-Ascend activation implementation:
  `vllm_ascend/ops/fused_moe/moe_mlp.py`
- vLLM-Omni diffusion MoE adapter:
  `vllm_omni/diffusion/layers/fused_moe.py`
- vLLM-Omni diffusion LoRA manager:
  `vllm_omni/diffusion/lora/manager.py`

### Feedback Period.

two weeks

### CC List.

@hsliuustc0106 @zengchuang-hw @Bounty-hunter @Gaohan123 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
